rust_src/src/multibyte.rs

```diff
@@ -348,13 +357,214 @@ impl LispStringRef {
         s.size
     }
 
-    /// Return width of STRING when displayed in the current buffer. Width is
-    /// measured by how many columns it occupies on the screen. When calculating
-    /// width of a multibyte character in STRING, only the base leading-code is
-    /// considered; the validity of the following bytes is not checked.  Tabs in
-    /// STRING are always taken to occupy `tab-width' columns.
+    /// Return the byte index corresponding to the given character index.
+    ///
+    /// Caches the ouput of the most recent calculation to make several calls on
+    /// the same string faster.
+    pub fn char_idx_to_byte(self, char_index: usize) -> usize {
+        let mut best_below = 0;
+        let mut best_below_byte = 0;
+        let mut best_above = self.len_chars() as usize;
+        let mut best_above_byte = self.len_bytes() as usize;
+
+        // Return if contents are unibyte.
+        if best_above == best_above_byte {
+            return char_index;
+        }
+
+        // Use cache operating on the cached string?
+        unsafe {
+            if LispObject::from(self) == string_index_cache {
+                // Check if character is before or after cached position
+                if charpos_cache < char_index {
+                    best_below = charpos_cache;
+                    best_below_byte = bytepos_cache;
+                } else {
+                    best_above = charpos_cache;
+                    best_above_byte = bytepos_cache
+                }
+            }
+        }
+
+        let byte_index = if char_index - best_below < best_above - char_index {
+            // Target is forwards.
+            let mut byte = best_below_byte;
+
+            while best_below < char_index {
+                byte += self.head_len(byte);
+                best_below += 1;
+            }
+
+            byte
+        } else {
+            // Target is backwards.
+            let mut byte = best_above_byte;
+
+            while best_above > char_index {
+                byte -= 1;
+                while !self.is_head_at(byte) {
+                    byte -= 1;
+                }
+                best_above -= 1;
+            }
+
+            byte
+        };
+
+        // Update cache
+        unsafe {
+            string_index_cache = self.into();
+            bytepos_cache = byte_index;
+            charpos_cache = char_index;
+        }
+        byte_index
+    }
+
```

Comment:
It certainly feels like there is a way to right a generic function here that use the char or byte type to do the right thing.

Reply:
It was quite a while ago so it's understandable if you don't remember, but I don't quite understand what you mean/meant, could you elaborate?

Reply:
I was referring to the if/else block that leads into this. It looks like something using the known type could help. But maybe not.

Reply:
I don't think so